File: src/reports/scheduler.ts
/**
 * DevFlow Monitor MCP - 보고서 스케줄러
 * 
 * 정기적인 보고서 생성을 관리하는 스케줄러입니다.
 */

import { EventEmitter } from 'eventemitter3';
import cron from 'node-cron';
import parser from 'cron-parser';
import { v4 as uuidv4 } from 'uuid';
import { Logger } from '../utils/logger.js';
import {
  ReportSchedule,
  SchedulePattern,
  ReportConfig,
  ReportResult,
  ReportEventType,
  ReportEvent
} from './types.js';
import { ReportEngine } from './report-engine.js';
import { ReportDelivery } from './delivery.js';
import { StorageManager } from '../storage/index.js';

/**
 * 스케줄러 설정
 */
export interface SchedulerConfig {
  /** 활성화 여부 */
  enabled?: boolean;
  
  /** 최대 동시 실행 수 */
  maxConcurrentJobs?: number;
  
  /** 기본 타임존 */
  defaultTimezone?: string;
  
  /** 스케줄 체크 간격 (밀리초) */
  checkInterval?: number;
  
  /** 오류 재시도 횟수 */
  retryAttempts?: number;
  
  /** 재시도 대기 시간 (밀리초) */
  retryDelay?: number;
}

/**
 * 스케줄 작업
 */
interface ScheduleJob {
  /** 작업 ID */
  id: string;
  
  /** 스케줄 */
  schedule: ReportSchedule;
  
  /** Cron 작업 */
  cronJob?: cron.ScheduledTask;
  
  /** 실행 중 여부 */
  running: boolean;
  
  /** 마지막 실행 결과 */
  lastResult?: ReportResult;
  
  /** 마지막 오류 */
  lastError?: string;
}

/**
 * 보고서 스케줄러
 */
export class ReportScheduler extends EventEmitter {
  private logger: Logger;
  private config: Required<SchedulerConfig>;
  private schedules: Map<string, ScheduleJob>;
  private reportEngine: ReportEngine;
  private reportDelivery: ReportDelivery;
  private storageManager: StorageManager;
  private checkTimer?: NodeJS.Timer;
  
  constructor(
    config: SchedulerConfig,
    reportEngine: ReportEngine,
    reportDelivery: ReportDelivery,
    storageManager: StorageManager
  ) {
    super();
    this.logger = new Logger('ReportScheduler');
    this.config = {
      enabled: true,
      maxConcurrentJobs: 5,
      defaultTimezone: 'UTC',
      checkInterval: 60000, // 1분
      retryAttempts: 3,
      retryDelay: 300000, // 5분
      ...config
    };
    
    this.schedules = new Map();
    this.reportEngine = reportEngine;
    this.reportDelivery = reportDelivery;
    this.storageManager = storageManager;
    
    if (this.config.enabled) {
      this.initialize();
    }
  }

  /**
   * 초기화
   */
  private async initialize(): Promise<void> {
    try {
      // 저장된 스케줄 로드
      await this.loadSchedules();
      
      // 정기 체크 시작
      this.startPeriodicCheck();
      
      this.logger.info('Report scheduler initialized', {
        schedulesCount: this.schedules.size
      });
    } catch (error) {
      this.logger.error('Failed to initialize scheduler', error);
    }
  }

  /**
   * 스케줄 생성
   */
  async createSchedule(
    name: string,
    reportConfig: ReportConfig,
    pattern: SchedulePattern,
    createdBy: string = 'system'
  ): Promise<ReportSchedule> {
    const schedule: ReportSchedule = {
      id: uuidv4(),
      name,
      enabled: true,
      reportConfig,
      schedule: pattern,
      createdBy,
      createdAt: Date.now(),
      updatedAt: Date.now()
    };
    
    // 타임존 설정
    if (!schedule.timezone) {
      schedule.timezone = this.config.defaultTimezone;
    }
    
    // 다음 실행 시간 계산
    schedule.nextRunAt = this.calculateNextRun(pattern);
    
    // 스케줄 작업 생성
    const job: ScheduleJob = {
      id: schedule.id,
      schedule,
      running: false
    };
    
    // Cron 작업 생성
    if (pattern.type === 'cron' && pattern.cron) {
      job.cronJob = this.createCronJob(schedule);
    }
    
    // 저장
    this.schedules.set(schedule.id, job);
    await this.saveSchedule(schedule);
    
    // 이벤트 발생
    this.emitScheduleEvent(ReportEventType.SCHEDULE_CREATED, schedule);
    
    this.logger.info('Schedule created', {
      scheduleId: schedule.id,
      name: schedule.name,
      pattern: pattern.type
    });
    
    return schedule;
  }

  /**
   * 스케줄 업데이트
   */
  async updateSchedule(
    scheduleId: string,
    updates: Partial<ReportSchedule>
  ): Promise<ReportSchedule | null> {
    const job = this.schedules.get(scheduleId);
    if (!job) {
      return null;
    }
    
    // 실행 중인 경우 거부
    if (job.running) {
      throw new Error('Cannot update schedule while it is running');
    }
    
    // 기존 Cron 작업 중지
    if (job.cronJob) {
      job.cronJob.stop();
      job.cronJob = undefined;
    }
    
    // 스케줄 업데이트
    const updatedSchedule: ReportSchedule = {
      ...job.schedule,
      ...updates,
      updatedAt: Date.now()
    };
    
    // 다음 실행 시간 재계산
    if (updates.schedule) {
      updatedSchedule.nextRunAt = this.calculateNextRun(updatedSchedule.schedule);
    }
    
    // 새 Cron 작업 생성
    if (updatedSchedule.enabled && updatedSchedule.schedule.type === 'cron' && updatedSchedule.schedule.cron) {
      job.cronJob = this.createCronJob(updatedSchedule);
    }
    
    // 업데이트
    job.schedule = updatedSchedule;
    await this.saveSchedule(updatedSchedule);
    
    // 이벤트 발생
    this.emitScheduleEvent(ReportEventType.SCHEDULE_UPDATED, updatedSchedule);
    
    this.logger.info('Schedule updated', { scheduleId });
    
    return updatedSchedule;
  }

  /**
   * 스케줄 삭제
   */
  async deleteSchedule(scheduleId: string): Promise<boolean> {
    const job = this.schedules.get(scheduleId);
    if (!job) {
      return false;
    }
    
    // 실행 중인 경우 거부
    if (job.running) {
      throw new Error('Cannot delete schedule while it is running');
    }
    
    // Cron 작업 중지
    if (job.cronJob) {
      job.cronJob.stop();
    }
    
    // 삭제
    this.schedules.delete(scheduleId);
    await this.removeSchedule(scheduleId);
    
    // 이벤트 발생
    this.emitScheduleEvent(ReportEventType.SCHEDULE_DELETED, job.schedule);
    
    this.logger.info('Schedule deleted', { scheduleId });
    
    return true;
  }

  /**
   * 스케줄 조회
   */
  getSchedule(scheduleId: string): ReportSchedule | null {
    const job = this.schedules.get(scheduleId);
    return job ? job.schedule : null;
  }

  /**
   * 모든 스케줄 조회
   */
  getAllSchedules(): ReportSchedule[] {
    return Array.from(this.schedules.values()).map(job => job.schedule);
  }

  /**
   * 스케줄 즉시 실행
   */
  async runScheduleNow(scheduleId: string): Promise<ReportResult> {
    const job = this.schedules.get(scheduleId);
    if (!job) {
      throw new Error('Schedule not found');
    }
    
    if (job.running) {
      throw new Error('Schedule is already running');
    }
    
    return this.executeSchedule(job);
  }

  /**
   * Cron 작업 생성
   */
  private createCronJob(schedule: ReportSchedule): cron.ScheduledTask {
    if (!schedule.schedule.cron) {
      throw new Error('Cron expression is required');
    }
    
    const task = cron.schedule(
      schedule.schedule.cron,
      async () => {
        const job = this.schedules.get(schedule.id);
        if (job && schedule.enabled) {
          try {
            await this.executeSchedule(job);
          } catch (error) {
            this.logger.error('Failed to execute scheduled report', {
              scheduleId: schedule.id,
              error
            });
          }
        }
      },
      {
        scheduled: true,
        timezone: schedule.timezone || this.config.defaultTimezone
      }
    );
    
    return task;
  }

  /**
   * 스케줄 실행
   */
  private async executeSchedule(job: ScheduleJob): Promise<ReportResult> {
    // 동시 실행 제한 확인
    const runningCount = Array.from(this.schedules.values()).filter(j => j.running).length;
    if (runningCount >= this.config.maxConcurrentJobs) {
      throw new Error('Maximum concurrent jobs reached');
    }
    
    job.running = true;
    const startTime = Date.now();
    
    try {
      this.logger.info('Executing scheduled report', {
        scheduleId: job.schedule.id,
        name: job.schedule.name
      });
      
      // 기간 계산
      const { periodStart, periodEnd } = this.calculateReportPeriod(job.schedule);
      
      // 보고서 생성
      const result = await this.reportEngine.generateReport(
        job.schedule.reportConfig,
        [], // 프로젝트 ID는 설정에서 가져올 수 있음
        periodStart,
        periodEnd
      );
      
      // 배포
      if (job.schedule.reportConfig.deliveryChannels?.length > 0) {
        const deliveryResults = await this.reportDelivery.deliver(
          result,
          job.schedule.reportConfig.deliveryChannels
        );
        result.deliveryResults = deliveryResults;
      }
      
      // 상태 업데이트
      job.lastResult = result;
      job.lastError = undefined;
      job.schedule.lastRunAt = startTime;
      job.schedule.nextRunAt = this.calculateNextRun(job.schedule.schedule);
      await this.saveSchedule(job.schedule);
      
      // 이벤트 발생
      this.emitScheduleEvent(ReportEventType.SCHEDULE_EXECUTED, job.schedule, result);
      
      this.logger.info('Scheduled report completed', {
        scheduleId: job.schedule.id,
        duration: Date.now() - startTime
      });
      
      return result;
    } catch (error) {
      // 오류 처리
      job.lastError = (error as Error).message;
      
      this.logger.error('Failed to execute scheduled report', {
        scheduleId: job.schedule.id,
        error
      });
      
      // 재시도 로직
      if (this.config.retryAttempts > 0) {
        this.scheduleRetry(job, 1);
      }
      
      throw error;
    } finally {
      job.running = false;
    }
  }

  /**
   * 재시도 스케줄링
   */
  private scheduleRetry(job: ScheduleJob, attempt: number): void {
    if (attempt > this.config.retryAttempts) {
      this.logger.error('Max retry attempts reached', {
        scheduleId: job.schedule.id
      });
      return;
    }
    
    setTimeout(async () => {
      try {
        await this.executeSchedule(job);
      } catch (error) {
        this.scheduleRetry(job, attempt + 1);
      }
    }, this.config.retryDelay * attempt);
  }

  /**
   * 다음 실행 시간 계산
   */
  private calculateNextRun(pattern: SchedulePattern): number {
    const now = new Date();
    
    switch (pattern.type) {
      case 'cron':
        if (pattern.cron) {
          const interval = parser.parseExpression(pattern.cron);
          return interval.next().getTime();
        }
        break;
        
      case 'interval':
        if (pattern.interval) {
          return now.getTime() + pattern.interval;
        }
        break;
        
      case 'daily':
        if (pattern.time) {
          const [hours, minutes] = pattern.time.split(':').map(Number);
          const next = new Date(now);
          next.setHours(hours, minutes, 0, 0);
          if (next.getTime() <= now.getTime()) {
            next.setDate(next.getDate() + 1);
          }
          return next.getTime();
        }
        break;
        
      case 'weekly':
        if (pattern.time && pattern.dayOfWeek !== undefined) {
          const [hours, minutes] = pattern.time.split(':').map(Number);
          const next = new Date(now);
          next.setHours(hours, minutes, 0, 0);
          
          // 다음 지정 요일 찾기
          const daysUntilNext = (pattern.dayOfWeek - now.getDay() + 7) % 7 || 7;
          next.setDate(next.getDate() + daysUntilNext);
          
          return next.getTime();
        }
        break;
        
      case 'monthly':
        if (pattern.time && pattern.dayOfMonth !== undefined) {
          const [hours, minutes] = pattern.time.split(':').map(Number);
          const next = new Date(now);
          next.setDate(pattern.dayOfMonth);
          next.setHours(hours, minutes, 0, 0);
          
          if (next.getTime() <= now.getTime()) {
            next.setMonth(next.getMonth() + 1);
          }
          
          return next.getTime();
        }
        break;
    }
    
    // 기본값: 1일 후
    return now.getTime() + 86400000;
  }

  /**
   * 보고 기간 계산
   */
  private calculateReportPeriod(schedule: ReportSchedule): {
    periodStart: number;
    periodEnd: number;
  } {
    const now = Date.now();
    let periodStart: number;
    let periodEnd: number = now;
    
    switch (schedule.reportConfig.type) {
      case 'daily':
        periodStart = now - 86400000; // 24시간
        break;
        
      case 'weekly':
        periodStart = now - 604800000; // 7일
        break;
        
      case 'monthly':
        periodStart = now - 2592000000; // 30일
        break;
        
      case 'quarterly':
        periodStart = now - 7776000000; // 90일
        break;
        
      default:
        periodStart = now - 86400000; // 기본 24시간
    }
    
    return { periodStart, periodEnd };
  }

  /**
   * 정기 체크 시작
   */
  private startPeriodicCheck(): void {
    this.checkTimer = setInterval(() => {
      this.checkSchedules();
    }, this.config.checkInterval);
  }

  /**
   * 스케줄 체크
   */
  private async checkSchedules(): Promise<void> {
    const now = Date.now();
    
    for (const job of this.schedules.values()) {
      if (!job.schedule.enabled || job.running) {
        continue;
      }
      
      // Cron이 아닌 스케줄 체크
      if (job.schedule.schedule.type !== 'cron' && job.schedule.nextRunAt) {
        if (job.schedule.nextRunAt <= now) {
          try {
            await this.executeSchedule(job);
          } catch (error) {
            this.logger.error('Failed to execute schedule', {
              scheduleId: job.schedule.id,
              error
            });
          }
        }
      }
    }
  }

  /**
   * 스케줄 로드
   */
  private async loadSchedules(): Promise<void> {
    try {
      // TODO: Implement schedule persistence
      const schedules: any[] = [];
      
      for (const data of schedules) {
        if (data.type === 'report_schedule') {
          const schedule = data.data as ReportSchedule;
          const job: ScheduleJob = {
            id: schedule.id,
            schedule,
            running: false
          };
          
          // Cron 작업 복원
          if (schedule.enabled && schedule.schedule.type === 'cron' && schedule.schedule.cron) {
            job.cronJob = this.createCronJob(schedule);
          }
          
          this.schedules.set(schedule.id, job);
        }
      }
      
      this.logger.info(`Loaded ${this.schedules.size} schedules`);
    } catch (error) {
      this.logger.error('Failed to load schedules', error);
    }
  }

  /**
   * 스케줄 저장
   */
  private async saveSchedule(schedule: ReportSchedule): Promise<void> {
    try {
      // TODO: Implement schedule persistence
      /*await this.storageManager.repositories.system.create({
        id: schedule.id,
        type: 'report_schedule',
        data: schedule,
        timestamp: Date.now()
      });*/
    } catch (error) {
      this.logger.error('Failed to save schedule', error);
    }
  }

  /**
   * 스케줄 제거
   */
  private async removeSchedule(scheduleId: string): Promise<void> {
    try {
      // 실제 구현에서는 ID로 삭제하는 메서드 필요
      this.logger.debug('Schedule removed from storage', { scheduleId });
    } catch (error) {
      this.logger.error('Failed to remove schedule', error);
    }
  }

  /**
   * 이벤트 발생
   */
  private emitScheduleEvent(
    type: ReportEventType,
    schedule: ReportSchedule,
    details?: any
  ): void {
    const event: ReportEvent = {
      type,
      reportId: schedule.id,
      timestamp: Date.now(),
      details: {
        schedule,
        ...details
      }
    };
    
    this.emit(type, event);
  }

  /**
   * 정리
   */
  async shutdown(): Promise<void> {
    // 타이머 중지
    if (this.checkTimer) {
      clearInterval(this.checkTimer);
    }
    
    // 모든 Cron 작업 중지
    for (const job of this.schedules.values()) {
      if (job.cronJob) {
        job.cronJob.stop();
      }
    }
    
    this.logger.info('Report scheduler shut down');
  }
}

File: python/src/devflow_monitor/reports/scheduler.py
"""
Report Scheduler.

Manages scheduled report generation using APScheduler.
"""

from __future__ import annotations

import asyncio
from datetime import datetime, timedelta
from typing import Any, Callable
from uuid import uuid4

from apscheduler.schedulers.asyncio import AsyncIOScheduler
from apscheduler.triggers.cron import CronTrigger
from apscheduler.triggers.interval import IntervalTrigger

from .delivery import ReportDelivery
from .report_engine import ReportEngine
from .types import (
    ReportConfig,
    ReportEventType,
    ReportResult,
    ReportSchedule,
    SchedulePattern,
)


class SchedulerConfig:
    """Scheduler configuration."""

    def __init__(
        self,
        enabled: bool = True,
        max_concurrent_jobs: int = 5,
        default_timezone: str = "UTC",
        check_interval: int = 60000,  # 1 minute in ms
        retry_attempts: int = 3,
        retry_delay: int = 300000,  # 5 minutes in ms
    ):
        """
        Initialize scheduler configuration.

        Args:
            enabled: Whether scheduler is enabled.
            max_concurrent_jobs: Maximum concurrent job executions.
            default_timezone: Default timezone for schedules.
            check_interval: Interval for checking schedules (ms).
            retry_attempts: Number of retry attempts on failure.
            retry_delay: Delay between retries (ms).
        """
        self.enabled = enabled
        self.max_concurrent_jobs = max_concurrent_jobs
        self.default_timezone = default_timezone
        self.check_interval = check_interval
        self.retry_attempts = retry_attempts
        self.retry_delay = retry_delay


class ScheduleJob:
    """Represents a scheduled job."""

    def __init__(
        self,
        id: str,
        schedule: ReportSchedule,
        running: bool = False,
    ):
        """Initialize schedule job."""
        self.id = id
        self.schedule = schedule
        self.running = running
        self.last_result: ReportResult | None = None
        self.last_error: str | None = None


class ReportScheduler:
    """
    Report scheduler.

    Manages scheduled report generation with support for
    cron expressions, intervals, and recurring patterns.
    """

    def __init__(
        self,
        config: SchedulerConfig | None = None,
        report_engine: ReportEngine | None = None,
        report_delivery: ReportDelivery | None = None,
    ):
        """
        Initialize the report scheduler.

        Args:
            config: Scheduler configuration.
            report_engine: Report engine instance.
            report_delivery: Report delivery instance.
        """
        self._config = config or SchedulerConfig()
        self._report_engine = report_engine
        self._report_delivery = report_delivery

        self._schedules: dict[str, ScheduleJob] = {}
        self._scheduler: AsyncIOScheduler | None = None
        self._listeners: dict[str, list[Callable]] = {}

        if self._config.enabled:
            self._initialize()

    def _initialize(self) -> None:
        """Initialize the scheduler."""
        self._scheduler = AsyncIOScheduler(timezone=self._config.default_timezone)
        self._scheduler.start()

    def on(self, event_type: str, handler: Callable) -> None:
        """Register an event listener."""
        if event_type not in self._listeners:
            self._listeners[event_type] = []
        self._listeners[event_type].append(handler)

    def emit(self, event_type: str, *args: Any) -> None:
        """Emit an event to listeners."""
        if event_type in self._listeners:
            for handler in self._listeners[event_type]:
                try:
                    handler(*args)
                except Exception:
                    pass

    async def create_schedule(
        self,
        name: str,
        report_config: ReportConfig,
        pattern: SchedulePattern,
        created_by: str = "system",
    ) -> ReportSchedule:
        """
        Create a new report schedule.

        Args:
            name: Schedule name.
            report_config: Report configuration.
            pattern: Schedule pattern.
            created_by: Creator identifier.

        Returns:
            Created report schedule.
        """
        schedule = ReportSchedule(
            id=str(uuid4()),
            name=name,
            enabled=True,
            report_config=report_config,
            schedule=pattern,
            timezone=self._config.default_timezone,
            created_by=created_by,
        )

        # Calculate next run time
        schedule.next_run_at = self._calculate_next_run(pattern)

        # Create job
        job = ScheduleJob(id=schedule.id, schedule=schedule)

        # Create scheduler job if cron pattern
        if pattern.type == "cron" and pattern.cron and self._scheduler:
            trigger = CronTrigger.from_crontab(
                pattern.cron,
                timezone=schedule.timezone,
            )
            self._scheduler.add_job(
                self._execute_schedule,
                trigger=trigger,
                id=schedule.id,
                args=[job],
            )

        # Store schedule
        self._schedules[schedule.id] = job

        # Emit event
        self.emit(ReportEventType.SCHEDULE_CREATED.value, {"schedule": schedule})

        return schedule

    async def update_schedule(
        self,
        schedule_id: str,
        updates: dict[str, Any],
    ) -> ReportSchedule | None:
        """
        Update an existing schedule.

        Args:
            schedule_id: Schedule ID to update.
            updates: Updates to apply.

        Returns:
            Updated schedule or None if not found.
        """
        job = self._schedules.get(schedule_id)
        if not job:
            return None

        if job.running:
            raise RuntimeError("Cannot update schedule while it is running")

        # Remove existing scheduler job
        if self._scheduler and self._scheduler.get_job(schedule_id):
            self._scheduler.remove_job(schedule_id)

        # Apply updates
        schedule_dict = job.schedule.model_dump()
        schedule_dict.update(updates)
        schedule_dict["updated_at"] = datetime.utcnow()

        # Recreate schedule
        updated_schedule = ReportSchedule(**schedule_dict)

        # Recalculate next run if pattern changed
        if "schedule" in updates:
            updated_schedule.next_run_at = self._calculate_next_run(updated_schedule.schedule)

        # Recreate scheduler job if enabled and cron
        if (
            updated_schedule.enabled
            and updated_schedule.schedule.type == "cron"
            and updated_schedule.schedule.cron
            and self._scheduler
        ):
            trigger = CronTrigger.from_crontab(
                updated_schedule.schedule.cron,
                timezone=updated_schedule.timezone,
            )
            self._scheduler.add_job(
                self._execute_schedule,
                trigger=trigger,
                id=schedule_id,
                args=[job],
            )

        # Update job
        job.schedule = updated_schedule

        # Emit event
        self.emit(ReportEventType.SCHEDULE_UPDATED.value, {"schedule": updated_schedule})

        return updated_schedule

    async def delete_schedule(self, schedule_id: str) -> bool:
        """
        Delete a schedule.

        Args:
            schedule_id: Schedule ID to delete.

        Returns:
            True if deleted, False if not found.
        """
        job = self._schedules.get(schedule_id)
        if not job:
            return False

        if job.running:
            raise RuntimeError("Cannot delete schedule while it is running")

        # Remove scheduler job
        if self._scheduler and self._scheduler.get_job(schedule_id):
            self._scheduler.remove_job(schedule_id)

        # Remove from schedules
        del self._schedules[schedule_id]

        # Emit event
        self.emit(ReportEventType.SCHEDULE_DELETED.value, {"schedule_id": schedule_id})

        return True

    def get_schedule(self, schedule_id: str) -> ReportSchedule | None:
        """
        Get a schedule by ID.

        Args:
            schedule_id: Schedule ID.

        Returns:
            Schedule or None if not found.
        """
        job = self._schedules.get(schedule_id)
        return job.schedule if job else None

    def get_all_schedules(self) -> list[ReportSchedule]:
        """
        Get all schedules.

        Returns:
            List of all schedules.
        """
        return [job.schedule for job in self._schedules.values()]

    async def run_schedule_now(self, schedule_id: str) -> ReportResult:
        """
        Execute a schedule immediately.

        Args:
            schedule_id: Schedule ID to run.

        Returns:
            Report generation result.

        Raises:
            ValueError: If schedule not found.
            RuntimeError: If schedule is already running.
        """
        job = self._schedules.get(schedule_id)
        if not job:
            raise ValueError("Schedule not found")

        if job.running:
            raise RuntimeError("Schedule is already running")

        return await self._execute_schedule(job)

    async def _execute_schedule(self, job: ScheduleJob) -> ReportResult:
        """Execute a scheduled report generation."""
        # Check concurrent job limit
        running_count = sum(1 for j in self._schedules.values() if j.running)
        if running_count >= self._config.max_concurrent_jobs:
            raise RuntimeError("Maximum concurrent jobs reached")

        job.running = True
        start_time = datetime.utcnow()

        try:
            # Calculate report period
            period_start, period_end = self._calculate_report_period(job.schedule)

            # Generate report
            if not self._report_engine:
                raise RuntimeError("Report engine not configured")

            result = await self._report_engine.generate_report(
                config=job.schedule.report_config,
                project_ids=[],
                period_start=period_start,
                period_end=period_end,
            )

            # Deliver report if channels configured
            if job.schedule.report_config.delivery_channels and self._report_delivery:
                delivery_results = await self._report_delivery.deliver(
                    result,
                    job.schedule.report_config.delivery_channels,
                )
                result.delivery_results = delivery_results

            # Update job state
            job.last_result = result
            job.last_error = None
            job.schedule.last_run_at = start_time
            job.schedule.next_run_at = self._calculate_next_run(job.schedule.schedule)

            # Emit event
            self.emit(
                ReportEventType.SCHEDULE_EXECUTED.value,
                {"schedule": job.schedule, "result": result},
            )

            return result

        except Exception as e:
            job.last_error = str(e)

            # Schedule retry
            if self._config.retry_attempts > 0:
                asyncio.create_task(self._schedule_retry(job, 1))

            raise

        finally:
            job.running = False

    async def _schedule_retry(self, job: ScheduleJob, attempt: int) -> None:
        """Schedule a retry attempt."""
        if attempt > self._config.retry_attempts:
            return

        # Wait before retry
        await asyncio.sleep(self._config.retry_delay * attempt / 1000)

        try:
            await self._execute_schedule(job)
        except Exception:
            await self._schedule_retry(job, attempt + 1)

    def _calculate_next_run(self, pattern: SchedulePattern) -> datetime:
        """Calculate the next run time based on pattern."""
        now = datetime.utcnow()

        if pattern.type == "cron" and pattern.cron:
            try:
                from croniter import croniter
                cron = croniter(pattern.cron, now)
                return cron.get_next(datetime)
            except ImportError:
                # Fallback: return next day
                return now + timedelta(days=1)

        elif pattern.type == "interval" and pattern.interval:
            return now + timedelta(milliseconds=pattern.interval)

        elif pattern.type == "daily" and pattern.time:
            time_parts = pattern.time.split(":")
            hours = int(time_parts[0]) if time_parts else 0
            minutes = int(time_parts[1]) if len(time_parts) > 1 else 0

            next_run = now.replace(hour=hours, minute=minutes, second=0, microsecond=0)
            if next_run <= now:
                next_run += timedelta(days=1)
            return next_run

        elif pattern.type == "weekly" and pattern.time and pattern.day_of_week is not None:
            time_parts = pattern.time.split(":")
            hours = int(time_parts[0]) if time_parts else 0
            minutes = int(time_parts[1]) if len(time_parts) > 1 else 0

            next_run = now.replace(hour=hours, minute=minutes, second=0, microsecond=0)
            days_until_next = (pattern.day_of_week - now.weekday() + 7) % 7 or 7
            next_run += timedelta(days=days_until_next)
            return next_run

        elif pattern.type == "monthly" and pattern.time and pattern.day_of_month is not None:
            time_parts = pattern.time.split(":")
            hours = int(time_parts[0]) if time_parts else 0
            minutes = int(time_parts[1]) if len(time_parts) > 1 else 0

            next_run = now.replace(
                day=pattern.day_of_month,
                hour=hours,
                minute=minutes,
                second=0,
                microsecond=0,
            )
            if next_run <= now:
                # Move to next month
                if now.month == 12:
                    next_run = next_run.replace(year=now.year + 1, month=1)
                else:
                    next_run = next_run.replace(month=now.month + 1)
            return next_run

        # Default: 1 day from now
        return now + timedelta(days=1)

    def _calculate_report_period(
        self,
        schedule: ReportSchedule,
    ) -> tuple[datetime, datetime]:
        """Calculate the report period based on schedule type."""
        now = datetime.utcnow()
        period_end = now

        report_type = schedule.report_config.type.value

        if report_type == "daily":
            period_start = now - timedelta(days=1)
        elif report_type == "weekly":
            period_start = now - timedelta(weeks=1)
        elif report_type == "monthly":
            period_start = now - timedelta(days=30)
        elif report_type == "quarterly":
            period_start = now - timedelta(days=90)
        else:
            period_start = now - timedelta(days=1)

        return period_start, period_end

    async def shutdown(self) -> None:
        """Shutdown the scheduler."""
        if self._scheduler:
            self._scheduler.shutdown(wait=True)

    def get_stats(self) -> dict[str, Any]:
        """
        Get scheduler statistics.

        Returns:
            Dictionary of scheduler statistics.
        """
        return {
            "enabled": self._config.enabled,
            "total_schedules": len(self._schedules),
            "running_jobs": sum(1 for j in self._schedules.values() if j.running),
            "failed_jobs": sum(1 for j in self._schedules.values() if j.last_error),
        }


# Singleton instance
_report_scheduler: ReportScheduler | None = None


def get_report_scheduler() -> ReportScheduler:
    """Get the singleton report scheduler instance."""
    global _report_scheduler
    if _report_scheduler is None:
        _report_scheduler = ReportScheduler()
    return _report_scheduler
